asdf/reference.py

```diff
@@ -12,7 +12,7 @@
 import numpy as np
 
 from . import generic_io, treeutil, util
-from .types import AsdfType
+from ._types import AsdfType
```

Comment:
Is `AsdfType` intended to be part of the private API? If so we might want to consider doing `from . import _types` instead, and accessing `AsdfType` via the `_type` namespace. This would apply in the other public modules that use `AsdfType`.

Reply:
AsdfType is currently part of the 'public' API at least with respect to the documentation: https://asdf.readthedocs.io/en/stable/api/asdf.types.AsdfType.html
This form of import was chosen to mimic the existing interface (where AsdfType is also exposed via `asdf.reference`).

Reply:
The developer docs say not

> This should not be used for user-defined extensions.

which indicates to me that this really shouldn't be public.

Reply:
I personally, don't think it should be public.

Reply:
I agree that AsdfType should not be public. This PR deprecates `asdf.types` so that we can eventually remove that submodule to start cleaning up the public API.

A follow up PR(s) will deprecate the public classes in `asdf.types` (AsdfType, ExtensionType) (I added these to the list of legacy api deprecations: https://github.com/asdf-format/asdf/issues/1333#issue-1558531069).

A more general clean up of our public API is very much called for but outside the scope of this work.

If making the changes you called our are required to get this PR approved I will make them. Are these required changes for this PR?

Reply:
> I agree that AsdfType should not be public. This PR deprecates `asdf.types` so that we can eventually remove that submodule to start cleaning up the public API.

> A follow up PR(s) will deprecate the public classes in `asdf.types` (AsdfType, ExtensionType) (I added these to the list of legacy api deprecations: [#1333 (comment)](https://github.com/asdf-format/asdf/issues/1333#issue-1558531069)).

I fail to understand how `AsdfType` is not effectively deprecated as part of the public API by simply deprecating the public module which contains it. Are you suggesting that `AsdfType` should be deprecated in the same fashion as `CustomType`, meaning depreciation of subclasses?

I guess this is fine if we plan on removing `AsdfType` from the public namespace of this public module before the `2.15` release.

Reply:
I made the changes you asked about.

To clarify, you consider this sufficient to deprecate `AsdfType`. Is the same true for `ExtensionType`?

Reply:
After grepping the codebase, I think `ExtensionType` should fall under the same category.